### PR TITLE
Support LDAP style oracle jdbc url

### LINKED ISSUE #19057: Ldap style oracle jdbc urls not supported for native
### Describe the bug

If I use a url `jdbc:oracle:thin:@ldap://...`, I get the following error at startup: ` java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory]`

### Expected behavior

Should work same as non ldap oracle jdbc url.

### Actual behavior

Here is the full exception:

```
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
2021-07-28 11:51:47,444 WARN  [io.agr.pool] (agroal-11) Datasource '<default>': IO Error: JNDI Package failure javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory] (CONNECTION_ID=yh0svrfuQK+prg+PVtb/mw==)
2021-07-28 11:51:47,445 WARN  [org.hib.eng.jdb.env.int.JdbcEnvironmentInitiator] (JPA Startup Thread: <default>) HHH000342: Could not obtain connection to query metadata: java.sql.SQLRecoverableException: IO Error: JNDI Package failure javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory] (CONNECTION_ID=yh0svrfuQK+prg+PVtb/mw==)
        at oracle.jdbc.driver.T4CConnection.handleLogonNetException(T4CConnection.java:870)
        at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:675)
        at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:1032)
        at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:90)
        at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:681)
        at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:602)
        at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:204)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:470)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:452)
        at java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:68)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(Thread.java:829)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: oracle.net.ns.NetException: JNDI Package failure javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory] (CONNECTION_ID=yh0svrfuQK+prg+PVtb/mw==)
        at oracle.net.jndi.JndiAttrs.initializeLDAPContext(JndiAttrs.java:290)
        at oracle.net.jndi.JndiAttrs.<init>(JndiAttrs.java:165)
        at oracle.net.resolver.AddrResolution.<init>(AddrResolution.java:334)
        at oracle.net.ns.NSProtocol.connect(NSProtocol.java:329)
        at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1967)
        at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:640)
        ... 14 more
Caused by: javax.naming.NoInitialContextException: Cannot instantiate class: com.sun.jndi.ldap.LdapCtxFactory [Root exception is java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory]
        at javax.naming.spi.NamingManager.getFactory(NamingManager.java:760)
        at javax.naming.spi.NamingManager.lambda$getInitialContext$1(NamingManager.java:722)
        at jdk.internal.loader.AbstractClassLoaderValue$Memoizer.get(AbstractClassLoaderValue.java:329)
        at jdk.internal.loader.AbstractClassLoaderValue.computeIfAbsent(AbstractClassLoaderValue.java:205)
        at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:722)
        at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
        at javax.naming.InitialContext.init(InitialContext.java:236)
        at javax.naming.InitialContext.<init>(InitialContext.java:208)
        at javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)
        at oracle.net.jndi.JndiAttrs.initializeLDAPContext(JndiAttrs.java:282)
        ... 19 more
Caused by: java.lang.ClassNotFoundException: com.sun.jndi.ldap.LdapCtxFactory
        at com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:64)
        at java.lang.Class.forName(DynamicHub.java:1308)
        at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:127)
        at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:133)
        at com.sun.naming.internal.VersionHelper.loadClass(VersionHelper.java:96)
        at javax.naming.spi.NamingManager.getFactory(NamingManager.java:757)
        ... 28 more
```

### How to Reproduce?

Create an application using the `quarkus-jdbc-oracle` driver.
Compile the application in native.
Run the application with a url of type: `quarkus.datasource.jdbc.url=jdbc:oracle:thin:@ldap://...`.
The application fails at startup.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

GraalVM 21.1.0 Java 11 CE (Java Version 11.0.11+8-jvmci-21.1-b05)

### Quarkus version or git rev

2.1.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Note that the application works with tns style url: `quarkus.datasource.jdbc.url=jdbc:oracle:thin:@(DESCRIPTION=(ENABLE=BROKEN)(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)...` (the address resolved by ldap).
So the issue is really about ldap access, not oracle.

### PR DESCRIPTION
**Pull request by vsevel**
Fixes https://github.com/quarkusio/quarkus/issues/19057

### CODE REVIEW

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java
Comment (gsmet) [inline extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: Ah I was thinking about it last night but didn't have time to play with it today. I'm not a big fan of always including the LDAP code but I have no idea how common this is.

I was wondering if maybe we should shoe in a new `DatabaseKind.ORACLE_LDAP` which would trigger this and behave exactly like an Oracle one.

Then in there, we would have a specific build step consuming JdbcDataSourceBuildItem and that would trigger the reflective registration if we have a datasource with this kind.

**Review by gsmet — CHANGES_REQUESTED** [review]
Let's agree on the best approach before merging this.

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java
Comment (vsevel) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: you are right this might not be so frequent. it is used extensively in my company btu we have tons of oracle instances. so ldap there is used for addressability (hiding the real connection details).
did not seem a big deal to add this class to reflection though, even if the client app is not using the ldap style. but I cannot measure the impact.
a new database kind seems to give that url format too much importance. another option would be to add an optional build time property `support_ldap_style_urls`.

**Review by Sanne — APPROVED** [review]
Excellent, thank you!

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java
Comment (Sanne) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: For as much as I'd want to fully avoid JNDI lookups across the board, even without this patch it would still work in JVM mode.

I'd rather maintain JVM ~ native as similar as possible and I agree it's no big deal to register a single class.

Comment (gsmet) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: Yeah so I ended up with this proposal because I don't think we have database-specific properties right now, unfortunately.

Measuring the impact is easy: build a native executable with and without this patch and compare the size of the executable.

Comment (Sanne) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: > another option would be to add an optional build time property support_ldap_style_urls

That's an interesting idea, but would imply adding some additional configuration which applies to Oracle only.

I think I'd rather merge this as-is, accepting that LDAP support already existed (although not functional in native in this case), and then possibly/optionally follow up with a general solution for all of Quarkus code to disable JNDI.

Which frankly I thought we already did, as GraalVM in early days wasn't able to compile core relying on JNDI - it seems clealry this is no longer failing.

Comment (Sanne) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: Ok I'll measure.

With the new reflection directive:
 - 77476 (76M)	./target/quarkus-integration-test-jpa-oracle-999-SNAPSHOT-runner

Without the reflection directive (LDAP style broken):
 - 76820 (76M)	./target/quarkus-integration-test-jpa-oracle-999-SNAPSHOT-runner

Comment (gsmet) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: Please let's have some data before merging this. It's not hard to collect and we will be able to take an educated decision. I will be happy to merge it if the cost is not prohibitive.

Comment (gsmet) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleReflections.java:26]: OK, perfect, let's merge then :). I was a bit concerned this type of things could bring half of Java with it including JAXB, AWT and Swing ;).

**Review by gsmet — APPROVED** [review]
Let's merge it. Cost is not high and it's not worth adding more complexity.

**Comment by vsevel** [comment]
> Let's merge it. Cost is not high and it's not worth adding more complexity.

@gsmet OK. I am running a few more tests to see if we need `methods` or `fields`or both.

**Comment by vsevel** [comment]
yes this works with `methods=false` and `fields=false`. let me do a quick change to save some extra bytes.

**Comment by vsevel** [comment]
@Sanne or  @gsmet ready for merge :)

**Comment by gsmet** [comment]
Thanks, I'll get this backported on Friday.

**Pull request merged**
